### PR TITLE
afinet: add option to set the maximum spoofed datagram size

### PR DESCRIPTION
**Pull request by bazsi**
Signed-off-by: Balazs Scheidler <balazs.scheidler@balabit.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by xtaran** [comment]
While I haven't checked if the same issue is present when using IPv6 nor if that afinet driver is IPv4 only, it looks to me as if this code is hardcoding IPv4-specific values:

```c
/* maximum ip datagram is 65535, minus IP + UDP header */
self->spoof_source_max_msglen = MAX(0, MIN(max_msglen, 65535 - 20 - 8));
```

So if not all we're looking at here is IPv4-only, this might cause some issues with UDP over IPv6.

File: modules/afsocket/afinet-dest.c [hunk 1]
```diff
@@ -98,6 +98,17 @@ afinet_dd_set_spoof_source(LogDriver *s, gboolean enable)
...
 }
 
+void
+afinet_dd_set_spoof_source_max_msglen(LogDriver *s, gint max_msglen)
+{
+#if SYSLOG_NG_ENABLE_SPOOF_SOURCE
+  AFInetDestDriver *self = (AFInetDestDriver *) s;
+
+  /* maximum ip datagram is 65535, minus IP + UDP header */
+  self->spoof_source_max_msglen = MAX(0, MIN(max_msglen, 65535 - 20 - 8));
```
[the remark this answers is not in the document]
Comment (Kokan) [reply modules/afsocket/afinet-dest.c]: The fact that the grammar ensures just a side effect imho, which an `afinet` setter should not depened on. But changing the `gint max_msglen` to `guint max_msglen` could provide a way to assume as much; end as it does not make sense to provide negative number the change makes sense here.

**Comment by bazsi** [comment]
You are right these are ipv4 specific values, but in reality these don't
really matter that much. In reality, anything that is more than a very few
fragments will be practically useless in production scenarios.

I can't remember the exact reasoning for the truncation, but my vague
recollection is that anything that starts to use IP fragments start to
cause message drops drastically. And in case a fragment is dropped, the
entire UDP payload will be dropped. And 1024 was the limit in RFC3164.

**Comment by xtaran** [comment]
> You are right these are ipv4 specific values, but in reality these don't really matter that much. In reality, anything that is more than a very few fragments will be practically useless in production scenarios.

True indeed.

My thought was though more going towards a potential buffer overflow or memory leak if this maximal allowed size is used in conjunction with IPv6 UDP packets… The current value seems to be the theoretically biggest usable size. (Which is probably be totally ok if using this with IPv6 packets doesn't cause any memory management issues — but I'm not versed enough with C to determine if this is the case.)

> I can't remember the exact reasoning for the truncation, […] And 1024 was the limit in RFC3164.

I suspected the latter as cause: The [old syslog RFC 3164 indeed limits the size to 1024 bytes](https://tools.ietf.org/html/rfc3164#section-4.1). And it also has a [requirement to make relayed syslog messages adhering to the RFC if they didn't do before](https://tools.ietf.org/html/rfc3164#section-4). These two rules together demands such a — from nowadays point of view unexpected — behaviour.

Anyway, thanks for caring and looking into this!

**Comment by gaborznagy** [comment]
Do we need to make this value configurable by users?
How about extending the max. value if we need a limit at all.
By adding an option to control this setting:
- the user should always take care of the set limit, and still can hit that limit
- it increases configuration complexity

As I see it after reading the linked RFC3164 parts, that we should only limit the length if we have to force RFC3164 format and not always.

**Comment by xtaran** [comment]
> As I see it after reading the linked RFC3164 parts, that we should only limit the length if we have to force RFC3164 format and not always.

Sounds ok for me, too.

File: modules/afsocket/afinet-dest.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply modules/afsocket/afinet-dest.c]: If we stick with having an option, I would like to request to have at least a warning here.

**Comment by bazsi** [comment]
I have now added a warning, although I don't see it really adds any value. I don't expect anyone hitting that. But with that this should be ok for merging.

File: modules/afsocket/afinet-dest.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (bazsi) [reply modules/afsocket/afinet-dest.c]: added

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**